### ISSUE TITLE
False-positive message "Avoid unused local variables" in anonymous class

### ISSUE THREAD

**Issue by Danila1985** [phpmd/phpmd#601]
There is false-positive message "Avoid unused local variables such as '$prop'" in anonymous class placed inside another class or function:
```
function getData($var)
{
    return new class($var)
    {
        private $prop;

        public function __construct($var)
        {
            $this->prop = $var;
        }

        public function getProp()
        {
            return $this->prop;
        }
    };
}

$data = getData('123');
echo $data->getProp();
```
Message: "Avoid unused local variables such as '$prop'."
Version: 2.6.0

**Comment by kylekatarnls** [comment]
Confirmed, it seems like anonymous class methods are evaluated as FunctionNode instead of MethodNode which breaks the link to `$prop`.

To be fixed in PDepend.

**Comment by kylekatarnls** [comment]
#783 should fix this bug, you can yet test ti requiring `"phpmd/phpmd": "dev-feature/issue-601-unused-property"`

**Comment by Danila1985** [comment]
"phpmd/phpmd": "dev-feature/issue-601-unused-property"  on clean project files failed with message:
> /vendor/bin/phpmd src,api.php,config text controversial,design,unusedcode
PHP Fatal error:  Uncaught Error: Call to undefined method PDepend\Source\AST\ASTArrayIndexExpression::getNode() in /vendor/phpmd/phpmd/src/main/php/PHPMD/Rule/AbstractLocalVariable.php:233
Stack trace:
#0 /vendor/phpmd/phpmd/src/main/php/PHPMD/Rule/AbstractLocalVariable.php(201): PHPMD\Rule\AbstractLocalVariable->prependMemberPrimaryPrefix('$accounts', Object(PHPMD\Node\ASTNode))
#1 /vendor/phpmd/phpmd/src/main/php/PHPMD/Rule/UnusedLocalVariable.php(146): PHPMD\Rule\AbstractLocalVariable->getVariableImage(Object(PHPMD\Node\ASTNode))
#2 /vendor/phpmd/phpmd/src/main/php/PHPMD/Rule/UnusedLocalVariable.php(94): PHPMD\Rule\UnusedLocalVariable->collectVariable(Object(PHPMD\Node\ASTNode))
#3 /vendor/phpmd/phpmd/src/main/php/PHPMD/Rule/UnusedLocalVariable.php(49): PHPMD\Rule\UnusedLocalVariable->collectVariables(Object(PHPMD\Node\MethodNode))
#4 /vendor/phpmd/phpmd/src/main/php/PHPMD/RuleSet.php(255): PHPMD\Rule\UnusedLocalVariable->apply(Object(PHPMD\Node\MethodNode))
#5 /vendor/phpmd/phpmd/src/main/php/PHPMD/Parser.php(259): PHPMD\ in /vendor/phpmd/phpmd/src/main/php/PHPMD/Rule/AbstractLocalVariable.php on line 233

on the issues example this:

>  /vendor/bin/phpmd tst.php text controversial,design,unusedcode
No node to visit provided for visitAnonymousClass.


version 2.8.2 hasn`t this problem, but old:

> Avoid unused local variables such as '$prop'.

**Comment by kylekatarnls** [comment]
Thanks for the feedback, I updated the branch, it should no longer throw this exception. (commit https://github.com/phpmd/phpmd/pull/783/commits/6c7a79706df18e0e39fab02a76955550ca0b6e19)

**Comment by Danila1985** [comment]
Thanks! All is fine

**Comment by kylekatarnls** [comment]
Thanks again for you help, this should be available in 2.9.0.
